src/history/historyTracker.ts

```diff
@@ -93,6 +93,50 @@ class HistoryStep {
     this.cursorStart = init.cursorStart || undefined;
     this.marks     = init.marks || [];
   }
+
+  /**
+   * merge collapses individual character changes into larger blocks of changes
+   */
+  public merge(): void {
+    if (this.changes.length < 2) {
+      return;
+    }
+    // create new changes list
+    var merged: DocumentChange[] = [];
+    // iterate through the changes, saving the first outside the loop
+    var iter = (this.changes[Symbol.iterator]());
+    var prev = iter.next().value;
+    for (const change of iter) {
+      if (prev.text.length === 0) {
+        // prev is eliminated, replace it with top of merged, or adopt change as prev
+        if (merged.length > 0) {
+          prev = merged.pop();
+        } else {
+          prev = change;
+          continue;
+        }
+      }
+      // merge logic. compares start & end() Positions to ensure this is the same location
+      if (prev.isAdd && change.isAdd && prev.end().isEqual(change.start)) {
+        // merge additions together
+        prev.text += change.text;
+      } else if (!prev.isAdd && !change.isAdd && change.end().isEqual(prev.start)) {
+        // merge deletions together, but in reverse so it still reads forward
+        change.text += prev.text;
+        prev = change;
+      } else if (prev.isAdd && !change.isAdd && prev.end().isEqual(change.end())) {
+        // collapse addition+deletion into addition
+        prev.text = prev.text.slice(0, -change.text.length);
+      } else {
+        // deletion+addition must be two separate DocumentChanges
```

Comment:
I want to note that there is a `TextEditor.replace` method that would be able to do both actions in one call, but the current DocumentChange does not support this (edge?)case right now and I didn't want to muck around too much.